### PR TITLE
[BUG] Patch doc and functionality for `is_minimal_d_separator`

### PR DESCRIPTION
**Pull request by adam2392**
Signed-off-by: Adam Li <adam2392@gmail.com>

Closes: #6247 
Fixes: #6216

I am opening this PR since the other one seems stalled.

### CODE REVIEW

**Review by dschult — COMMENTED** [review]
My concern with both of these PRs is that I'm not convinced that this is the best solution to the ambiguity. It seems like rewriting the doc_string to make it clear what each function does will allow users to choose effectively and to avoid confusion.

To try changing the doc_string, we need to rewrite the paragraph description to make it clear that the `is_minimal_d_separator` function checks whether the inputted d-separator set is minimal. Perhaps in addition to calling the `z` input "a set that is a valid d_separator", there should be a second paragraph with a sentence:  

```
Note: This function does not check whether `z` is a d_separator. Use the function `d_separated` together with this function (as shown in the examples below) to check that `z` is both minimal and a d_separator.
```

Then we should add an **Examples** section with a DiGraph and set `z` where we check that `z` is a d_separator and then check that it is minimal.

Thoughts?

I've also got a minor suggestion about wording in the doc_string. And I can't find a test of the case when there is a valid d_separator that is not minimal. I that an easy thing to add?

**Comment by adam2392** [comment]
> My concern with both of these PRs is that I'm not convinced that this is the best solution to the ambiguity. It seems like rewriting the doc_string to make it clear what each function does will allow users to choose effectively and to avoid confusion.
> Then we should add an Examples section with a DiGraph and set z where we check that z is a d_separator and then check that it is minimal.

Just to clarify, it seems like there are three possible suggestions/solns:

1. modify doc string to clarify that checking for d-separation is the job of the user (or that the function will check for d-separation)
2. implement additional check for d-separation
3. implement both, but also allow the user to turn off the check for d-separation(?) possibly for efficiency sake.

I'm in favor of 3 since that seems to handle both you and Ross's suggestions. I can also add an example.

**Comment by adam2392** [comment]
Okay, I went ahead and added:

- an option to turn on/off the check for d-separation in `is_minimal_d_separator`
- clarification in the docstring
- example in the docstring
- an extra test case of when minimal-d-separator is met and not met

Lmk if that works?

File: networkx/algorithms/d_separation.py [hunk 1]
```diff
@@ -312,12 +316,36 @@ def is_minimal_d_separator(G, u, v, z):
...
+    False
+    >>>
+    >>> # if we already know that z is a d-separator, we can turn off the check
+    >>> nx.is_minimal_d_separator(G, 0, 2, {1}, False)
+    True
+    >>> # if we do not know that z is a d-separator, then we can first check it ourselves
+    >>> is_d_sep = nx.d_separated(G, {0}, {4}, {1})
+    >>> is_minimal = nx.is_minimal_d_separator(G, 0, 2, {1}, False)
+    >>> print(is_d_sep and is_minimal)
+    True
```
Comment (dschult) [inline networkx/algorithms/d_separation.py:348]: I think the code for checking both d_separator and minimal is simpler than what appears here. That's why I'm thinking it might be better to avoid the keyword argument. It would be different if the algorithm for minimal and for d_separated were more similar. But it feels uncomfortable to try to join two functions that are really asking different questions.
```python
# this seems straightforward to me
if nx.d_separated(G, {u}, {v}, z) and nx.is_minimal_d_separator(G, u, v, z):
    print("This d_separator is minimal")
```
Would a better name for `is_minimal_d_separator` solve the problem?

Or... Is there some way to use the algorithm for "minimal" and as part of that algorithm check that z is a d_separator?

Comment (adam2392) [reply networkx/algorithms/d_separation.py:348]: > Or... Is there some way to use the algorithm for "minimal" and as part of that algorithm check that z is a d_separator?

Unfortunately, not from what I see. IIUC, the implicit assumption for `is_minimal_d_separator` is that a set is already a d-separator. But Algorithm 1 seems to have the check of d-separator internally: https://ftp.cs.ucla.edu/pub/stat_ser/r254.pdf. E.g. in Theorem 5 d-separation is assumed apriori. But Theorem 4 has a check for d-separation explicitly.

We could implement algo 1 instead?

Alternatively, perhaps we should rename the overall function to `is_d_separator_minimal`, so that way it implies that one already has a d-separator. 

In summary:
1. I could rename the function
2. I could remove the `check_d_sep` kwarg
3. Or do something else.

WDYT?

Comment (rossbar) [reply networkx/algorithms/d_separation.py:348]: FWIW my [original opinion](https://github.com/networkx/networkx/issues/6216#issuecomment-1335938007) hasn't changed:
 1. I'm -1 on a keyword argument. Given the function name, most users will assume that the function checks that a d-separator is both valid **and** minimal. Adding a keyword argument requires the users to study the API in greater detail (i.e. look at each parameter in detail and read the docstring) to find out this isn't the case, and thus is poor API design IMO.
 2. Though it might help the situation, I'm also -1 on adding a new function name. We can't get rid of `is_minimal_d_separator` without a deprecation period, and IMO it'd be too confusing to have **both** `is_minimal_d_separator` and `is_d_separator_minimal`, even if there is a deprecation warning indicating which should be used.

Given the above, I think the course of action that makes the most sense remains to add the `d_separated` check internally to to the `is_minimal_d_separator` function. Again, see my original comment for more detailed explanation of the rationale, but given the API already in place, this is the course of action that is minimally disruptive. The cost of course is that `is_minimal_d_separator` is unnecessarily expensive in the case where a user already knows the nodes are `d_separated`. To me, a clear & simple user interface is worth the performance hit.

Comment (dschult) [reply networkx/algorithms/d_separation.py:348]: I think I'm spinning our wheels in this discussion. Sorry about that. And Thanks @adam2392 and @rossbar for making it more clear to me what's going on.

I'm now OK with the original proposal without the keyword and with always verifying that `z` is a d_separator.
Sorry @adam2392 but could you remove the keyword again and get it back to the API you originally had?
I do like the additional tests you added and doc_string changes other than for the keyword...
:}

Thanks!

Comment (adam2392) [reply networkx/algorithms/d_separation.py]: Done in https://github.com/networkx/networkx/pull/6427/commits/5bd573a6ead6c4482f812616790b68effb00ccfe!

File: networkx/algorithms/d_separation.py [hunk 2]
```diff
@@ -312,12 +316,27 @@ def is_minimal_d_separator(G, u, v, z):
     v : node
         A node in the graph.
     z : Set of nodes
-        The set of nodes to check if it is a minimal d-separating set.
+        The set of nodes to check if it is a minimal d-separating set. A call
+        to :func:`d_separated` is called to verify that `z` is in fact a
+        d-separator is always made.
```
Comment (dschult) [inline networkx/algorithms/d_separation.py:321]:
```suggestion
        The set of nodes to check if it is a minimal d-separating set.
        The function :func:`d_separated` is called inside this function
        to verify that `z` is in fact a d-separator.
```

File: networkx/algorithms/d_separation.py [hunk 3]
```diff
@@ -312,12 +316,27 @@ def is_minimal_d_separator(G, u, v, z):
...
+
+    Examples
+    --------
+    >>> G = nx.path_graph([0, 1, 2, 3], create_using=nx.DiGraph)
+    >>> G.add_node(4)
+    >>> # since {1} is the minimal d-separator, this will return False
+    >>> nx.is_minimal_d_separator(G, 0, 2, {1,3,4})
+    False
+    >>> nx.is_minimal_d_separator(G, 0, 2, {1})
+    True
```
Comment (dschult) [inline networkx/algorithms/d_separation.py:336]:
```suggestion
    >>> nx.is_minimal_d_separator(G, 0, 2, {1})
    True
    >>> # since {1} is the minimal d-separator, {1, 3, 4} is not minimal
    >>> nx.is_minimal_d_separator(G, 0, 2, {1, 3, 4})
    False
```
Put True result before comment, replace pronoun "this" and add spaces to the set.

File: networkx/algorithms/d_separation.py [hunk 4]
```diff
@@ -312,12 +316,27 @@ def is_minimal_d_separator(G, u, v, z):
...
+    >>> # alternatively, if we simply want to check that {1,3,4} is a d-separator
+    >>> nx.d_separated(G, {0}, {4}, {1,3,4})
```
Comment (dschult) [inline networkx/algorithms/d_separation.py:338]:
```suggestion
    >>> # alternatively, if we only want to check that {1, 3, 4} is a d-separator
    >>> nx.d_separated(G, {0}, {4}, {1, 3, 4})
```

File: networkx/algorithms/d_separation.py [hunk 5]
```diff
@@ -359,6 +378,9 @@ def is_minimal_d_separator(G, u, v, z):
     if not nx.is_directed_acyclic_graph(G):
         raise nx.NetworkXError("graph should be directed acyclic")
 
+    if not nx.d_separated(G, {u}, {v}, z):
+        return False
+
```
Comment (dschult) [inline networkx/algorithms/d_separation.py:383]: It looks like this call to `d_separated` calls `nx.is_directed_acyclic_graph` so we should remove the previous error check.

Also: The `union_uv` construction and error check on lines 384-389, are done inside the `d_separated` call. So those should be deleted too.

**Review by dschult — COMMENTED** [review]
OK... finally down to the code. :)
I have some word-smithing (which are just suggestions... feel free to reword, decide not to do, etc) and two error checks that are now duplicated because they are done in `d_separated`. 

Thanks for this.

File: networkx/algorithms/d_separation.py (hunk 5 above)
Comment (adam2392) [reply networkx/algorithms/d_separation.py]: Good catch. Removed.

File: networkx/algorithms/d_separation.py [hunk 6]
```diff
@@ -356,15 +375,12 @@ def is_minimal_d_separator(G, u, v, z):
 
     https://en.wikipedia.org/wiki/Bayesian_network#d-separation
     """
-    if not nx.is_directed_acyclic_graph(G):
-        raise nx.NetworkXError("graph should be directed acyclic")
+    if not nx.d_separated(G, {u}, {v}, z):
+        return False
 
     union_uv = {u, v}
     union_uv.update(z)
```
Comment (dschult) [inline networkx/algorithms/d_separation.py:382]: I think union_uv is no longer used. So we can remove this construction.

File: networkx/algorithms/tests/test_d_separation.py [hunk 7]
```diff
@@ -154,9 +154,9 @@ def test_cyclic_graphs_raise_error():
     with pytest.raises(nx.NetworkXError):
         nx.d_separated(g, {0}, {1}, {2})
     with pytest.raises(nx.NetworkXError):
-        nx.minimal_d_separator(g, {0}, {1})
+        nx.minimal_d_separator(g, 0, 1)
```
Comment (dschult) [inline networkx/algorithms/tests/test_d_separation.py:157]: Ooo...  That's a good catch of a sneaky test error. The inputs aren't nodes, but it was raising an error before it ever checked if they were nodes. Still doing that, bu at least the input nodes are now valid nodes. :}

File: networkx/algorithms/d_separation.py (hunk 6 above)
Comment (adam2392) [reply networkx/algorithms/d_separation.py]: Done

File: networkx/algorithms/tests/test_d_separation.py (hunk 7 above)
Comment (adam2392) [reply networkx/algorithms/tests/test_d_separation.py:157]: Yep and at least the docstring is clear that the algorithm does not work for a set of nodes, so that is good

**Review by dschult — APPROVED** [review]
Thanks!

**Review by jarrodmillman — APPROVED** [review]
